File: leetcode/950-reveal-cards-in-increasing-order.js
// https://leetcode.com/problems/reveal-cards-in-increasing-order/

/**
 * @param {number[]} deck
 * @return {number[]}
 */
var deckRevealedIncreasing = function(deck) {
  deck.sort((a, b) => (a - b));
  const queue = deck.map((val, index) => index);
  const output = new Array(deck.length);
  for (const card of deck) {
    output[queue.shift()] = card;
    if (queue.length > 0) queue.push(queue.shift());
  }

  return output;
};

File: leetcode/950-reveal-cards-in-increasing-order.py
# https://leetcode.com/problems/reveal-cards-in-increasing-order/

from collections import deque
from typing import List

class Solution:
    def deckRevealedIncreasing(self, deck: List[int]) -> List[int]:
        deck.sort()
        queue = deque(range(len(deck)))
        output = [None] * len(deck)
        for card in deck:
            output[queue.popleft()] = card
            if queue: queue.append(queue.popleft())
        
        return output
